src/api.rs

```diff
@@ -64,4 +67,6 @@ pub trait Introspectable {
     fn resume(&mut self) -> Result<(), Box<dyn Error>> {
         unimplemented!();
     }
+
+    fn get_driver_type(&self) -> DriverType;
```

Comment:
@rageagainsthepc can you explain why you had to introduce a new API here, it's hard to follow for me ?
I would like to avoid changing the core of the Rust library just to provide a C compatibility layer.

Reply:
Well, if we initialize the driver with None, we don't know the type of the driver in advance. But we need to know since we are casting back and forth between void* and the actual type of the driver. As far as I know it's not possible to cast directly from void* to a trait type. There are probably different ways to approach this, but this one seemed pretty straight forward to me. I'm open to alternatives though ;)

Reply:
I think the issue was that pointers to traits are fat pointers. Maybe there is way to pass fat pointers to C code and cast them back to traits afterwards. I'll look into that.

Reply:
There exists a std::raw::TraitObject but it's marked as unstable because its representation may change in the future https://github.com/rust-lang/rust/issues/27751 . Also cbindgen reports TraitObject types as unsupported. In theory we could split the fat pointer into two separate pointers, store them in a struct and transmute the memory back to an Introspectable, but imho that sounds super ugly and unstable.

Reply:
On the other hand, we could just loose the support for passing `None` instead of the DriverType ;)

Reply:
I've found the crate downcast-rs. With this crate you are able to check a crate for its concrete type.

Reply:
Okay, thanks for the update.

I'm okay to merge this as long as you add a comment that this API has been introduced for the solve purpose of the C interoperability, and that it's a workaround, to be deprecated in the future when new better solutions will emerge to deal with this, either from Rust unstable, or simply by switching to dynamic library loading.